crates/primitives/src/checkpoints.rs

```diff
@@ -121,6 +123,88 @@ pub struct StorageHashingCheckpoint {
     pub to: u64,
 }
 
+/// Saves the progress of a stage.
+#[main_codec]
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct StageCheckpoint {
+    /// Stage progress measured in block number it reached.
+    pub block_number: BlockNumber,
+    /// Stage-specific checkpoint. None if stage uses only block-based checkpoints.
+    pub stage_checkpoint: Option<StageUnitCheckpoint>,
+}
+
+impl StageCheckpoint {
+    /// Creates a new [`StageCheckpoint`] with only `block_number` set.
+    pub fn new_with_block_number(block_number: BlockNumber) -> Self {
+        Self { block_number, ..Default::default() }
+    }
+}
+
+// TODO(alexey): ideally, we'd want to display block number + stage-specific metric (if available)
+//  in places like logs or traces
+impl Display for StageCheckpoint {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.block_number, f)
+    }
+}
+
+// TODO(alexey): add a merkle checkpoint. Currently it's hard because [`MerkleCheckpoint`]
+//  is not a Copy type.
+/// Stage-specific checkpoint metrics.
+#[derive_arbitrary(compact)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
+pub enum StageUnitCheckpoint {
```

Comment:
why does unit have to be `Copy`?

Reply:
due to `StageCheckpoint` being `Copy`, which in turn due to `ExecInput`/`UnwindInput` which we copy sometimes in the pipeline. Plan to fix that in the future PR to not make this one even bigger.